### DISCUSSION TITLE
NFS Sharedata (other server / Alt 2) does not generate / transfer sharedata files to the NAS.

### DISCUSSION THREAD

**Discussion by Sagnasty** [rern/rAudio#1704] [Bug Reports]
Currently running rAudio 20240706

I attempted to create a clean build of the Shareddata files on my NAS using the following procedure:

1. Deleted all pre-existing shared data files from my shareddata directory on my NAS.
2. Via systems page, I entered the server IP and path information and pressed OK.
3. System successfully created the additional mount as viewed from the network storage area.

I then noticed that the shareddata file structure was not created on my NAS as it had in the past (no files and no directories).   Also, clicking off the systems page to the library page generated the following error message:
![image](https://github.com/rern/rAudio/assets/76575081/eefac8cf-0580-453d-804a-d095f3d35a2e)
I then clicked OK to revert back to using the local data

I cannot say that this issue was introduced in 20240706 as I have not updated my shareddata in quite some time (20240310 release was the last known time I build the shareddata structure on my NAS).

I also tried this on another machine with exactly the same results.

Jim

3 x Raspberry Pi 3B+ Rev 1.3 > HiFiBerry DAC+ Pro
1 x Raspberry Pi 3B Rev 1.2 > HiFiBerry DAC+ Pro
Synology DS718+

**Comment by ceejayemm** [comment]
Same here, built new system using RPi3B+/i20240616 and could not create Sharedata, using CIFS nor NFS, so had to resort to a local installation.

New share gets mounted as shown in /etc/fstab  (password blanked after cut/paste):
:
PARTUUID=cb90bf37-01                       /boot                   vfat  defaults,noatime                                                                 >
PARTUUID=cb90bf37-02                       /                       ext4  defaults,noatime                                                                 >
192.168.86.250:/export/music/iTunes/Music  /mnt/MPD/NAS/NAS_Music  nfs   defaults,noauto,bg,soft,timeo=5                                                  >
//192.168.86.250/radb                      /mnt/MPD/NAS/data1      cifs  noauto,username=raudio,password=XXXXXXXXX,uid=45,gid=45,iocharset=utf8,vers=3.1.1>

but no data is transferred.

Running 'journalctl -f' during process shows (password blanked after cut/paste):

Jul 07 10:16:36 RAOffice sudo[10731]:     http : PWD=/srv/http ; USER=root ; COMMAND=/srv/http/bash/settings/system-mount.sh 'cmd#012cifs#012data1#012192.168.86.250#012radb#012raudio#012XXXXXXXX#012vers=3.1.1#012true#012CMD PROTOCOL NAME IP SHARE USR PASSWORD OPTIONS SHAREDDATA'
Jul 07 10:16:36 RAOffice sudo[10731]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=33)
Jul 07 10:16:36 RAOffice systemd[1]: mnt-MPD-NAS-data1.mount: Deactivated successfully.
Jul 07 10:16:36 RAOffice systemd[1]: Reloading requested from client PID 10747 ('systemctl') (unit php-fpm.service)...
Jul 07 10:16:36 RAOffice systemd[1]: Reloading...
Jul 07 10:16:36 RAOffice systemd[1]: Binding to IPv6 address not available since kernel does not support IPv6.
Jul 07 10:16:36 RAOffice systemd[1]: Binding to IPv6 address not available since kernel does not support IPv6.
Jul 07 10:16:37 RAOffice systemd[1]: Reloading finished in 809 ms.
Jul 07 10:16:37 RAOffice kernel: CIFS: enabling forceuid mount option implicitly because uid= option is specified
Jul 07 10:16:37 RAOffice kernel: CIFS: enabling forcegid mount option implicitly because gid= option is specified
Jul 07 10:16:37 RAOffice kernel: CIFS: Attempting to mount //192.168.86.250/radb
Jul 07 10:16:38 RAOffice systemd[1]: Stopping devmon Service...
Jul 07 10:16:38 RAOffice devmon[10476]: devmon: stopped
Jul 07 10:16:38 RAOffice systemd[1]: devmon@http.service: Deactivated successfully.
Jul 07 10:16:38 RAOffice systemd[1]: Stopped devmon Service.
Jul 07 10:16:38 RAOffice systemd[1]: Started devmon Service.
Jul 07 10:16:38 RAOffice systemd[1]: Stopping MPD idle...
Jul 07 10:16:38 RAOffice systemd[1]: mpdidle.service: Deactivated successfully.
Jul 07 10:16:38 RAOffice systemd[1]: Stopped MPD idle.
Jul 07 10:16:38 RAOffice systemd[1]: Stopping Music Player Daemon...
Jul 07 10:16:38 RAOffice systemd[1]: mpd.service: Deactivated successfully.
Jul 07 10:16:38 RAOffice systemd[1]: Stopped Music Player Daemon.
Jul 07 10:16:38 RAOffice systemd[1]: mpd.service: Consumed 1.089s CPU time.
Jul 07 10:16:38 RAOffice systemd[1]: Starting Music Player Daemon...
Jul 07 10:16:39 RAOffice mpd[10852]: exception: Failed to access /srv/http/data/playlists: No such file or directory
Jul 07 10:16:39 RAOffice mpd[10852]: decoder: Decoder plugin 'wildmidi' is unavailable: configuration file does not exist: /etc/timidity/timidity.cfg
Jul 07 10:16:39 RAOffice mpd[10852]: exception: Failed to open '/srv/http/data/mpd/mpd.db': No such file or directory
Jul 07 10:16:39 RAOffice mpd[10852]: exception: Failed to open database plugin; On parent directory of db file; Failed to access /srv/http/data/mpd: No such file or directory
Jul 07 10:16:39 RAOffice systemd[1]: mpd.service: Main process exited, code=exited, status=1/FAILURE
Jul 07 10:16:39 RAOffice systemd[1]: mpd.service: Failed with result 'exit-code'.
Jul 07 10:16:39 RAOffice systemd[1]: Failed to start Music Player Daemon.
Jul 07 10:16:39 RAOffice systemd[1]: Dependency failed for MPD idle.
Jul 07 10:16:39 RAOffice systemd[1]: mpdidle.service: Job mpdidle.service/start failed with result 'dependency'.
Jul 07 10:16:40 RAOffice systemd[1]: Starting Time & Date Service...
Jul 07 10:16:40 RAOffice systemd[1]: Started Time & Date Service.
Jul 07 10:16:41 RAOffice sudo[10731]: pam_unix(sudo:session): session closed for user root
Jul 07 10:17:10 RAOffice systemd[1]: systemd-timedated.service: Deactivated successfully.

Chris

**Reply by rern** [reply]
Hi,

Please try: `sed -i 's/infoval.NAME =/! shareddata \&\& &/' /srv/http/assets/js/system.js`

**Reply by ceejayemm** [reply]
@rern 
Thanks, that seems to at least make the Shareddata area available and it gets populated from the first system connected to it.  I will do some more testing but so far looks good.

Thanks

Chris

**Comment by Sagnasty** [comment]
Hi Rern,
I can also confirm that shared data was created / transferred to my NAS via NFS.
Thanks, Jim

**Comment by ceejayemm** [comment]
@rern 
As noted above the first system seemed to work ok once the suggested edit was applied.  I have since tried to set up a second RPi (3A+) to access the same NAS CIFS share.  I applied the edit but I still get the same error as originally reported and the 3A reverts to local storage.

Something also seems to be amiss when using the rAudio provided NFS server for shared data.  The first system (RPi 3B+) creates the NFS share but the RPi3A fails to attach to it - might be a coincidence or related ?  Do you want me to raise a separate issue for this ?

Chris

**Comment by ceejayemm** [comment]
@ rern
In regard to the rAudio provided Shared Data problem. It would seem that the system setup as the server is providing an NFS share but the second machine (client) is not mounting the share correctly.  When Shared Data is enabled on the client, it  seems to create a folder:

```
rAudio-1:~  #  ls -lrt /mnt/MPD/NAS
total 8
drwxr-xr-x 2 mpd  audio 4096 Jul  1 09:06 NAS_Music
drwxrwxrwx 8 http http  4096 Jul 11 07:55 data

```
but the share is being mounted (via FSTAB) to:

```
PARTUUID=d062bfbc-01                       /boot                   vfat  defaults,noatime                 0  0
PARTUUID=d062bfbc-02                       /                       ext4  defaults,noatime                 0  0
192.168.86.250:/export/music/iTunes/Music  /mnt/MPD/NAS/NAS_Music  nfs   defaults,noauto,bg,soft,timeo=5  0  0
192.168.86.31:/mnt/MPD/NAS                 /mnt/MPD/NAS            nfs   defaults,noauto,bg,soft,timeo=5  0  0

```
and:

```
rAudio-1:~  #  df -h
Filesystem                                 Size  Used Avail Use% Mounted on
/dev/root                                  7.1G  3.0G  3.8G  45% /
devtmpfs                                   142M     0  142M   0% /dev
tmpfs                                      176M  132K  176M   1% /dev/shm
tmpfs                                       71M  656K   70M   1% /run
tmpfs                                      176M  4.0K  176M   1% /tmp
/dev/mmcblk0p1                             197M   49M  149M  25% /boot
192.168.86.250:/export/music/iTunes/Music  6.9G  3.1G  3.5G  48% /mnt/MPD/NAS/NAS_Music
192.168.86.31:/mnt/MPD/NAS                 6.9G  3.1G  3.5G  48% /mnt/MPD/NAS
```

Should the NFS share be mounted to /mnt/MPD/NAS/data rather than /mnt/MPD/NAS ?

Also the System / Data and Settings / Shared Data switch always seems to return to the disabled position after an attempt to set up the shared data setting has been made.

Chris

**Reply by rern** [reply]
Please try the UPDATE.

**Comment by Sagnasty** [comment]
Hi - Out of town until early next week.

**Comment by ceejayemm** [comment]
@rern
I have applied the latest UPDATE branch to the server and client, the update process, on the client only, finished with:

```
chown: changing ownership of '/srv/http/data/mpd': Operation not permitted
chown: changing ownership of '/srv/http/data/playlists': Operation not permitted
chown: changing ownership of '/srv/http/data/mpd/mpd.db': Operation not permitted
```
I have tried out the changes and, having removed and reinstated the Shared Data settings on the client, everything now seems to work as I would expect.  I will continue testing and report any findings.

Thanks

Chris

**Comment by ceejayemm** [comment]
Just applied latest UPDATE branch to client system but still get:

```
chown: changing ownership of '/srv/http/data/mpd/mpd.db': Operation not permitted
```


Chris

**Comment by Sagnasty** [comment]
Hi,

Today I applied the UPDATE branch to my server (Raspberry Pi 3B+ Rev 1.3 > HiFiBerry DAC+ Pro) and client (Raspberry Pi 3B Rev 1.2) machines via NFS.

Enabling shareddata on the server machine successfully created the shareddata on my NAS

Enabling sharedata on the client machine successfully established a storage mount to the shareddata path on the NAS and correctly displayed the contents.


On the client machine local library mount, I use an IP address which is provided by DDNS on the NAS for both local and external network use (same as I always have).   When I use this machine on my local network, the IP address is correctly translated by the router with no issues.
The problem is, during the shareddata mount creation process, two shareddata mounts are being created: a local IP NAS address (x.x.x.4) mount and the DDNS IP NAS address (x.x.x.251) mount.   The shareddata works fine at this point.

![image](https://github.com/user-attachments/assets/6630f237-962e-4ed1-a4e5-7cf422569a94)



However, when DISabling the client sharedata, the storage mount for local data was also removed in addition to removing the two shareddata mounts.

![image](https://github.com/user-attachments/assets/278bd39d-ab8a-4aab-b8d1-5bd9cdefb9d3)


Restoring a backup was successful to regain my local library mount.

![image](https://github.com/user-attachments/assets/be264009-8bed-4710-9468-360bb5b1afcf)


Jim

**Comment by Sagnasty** [comment]
Hi,

Same behavior as described above after applying rAudio 20240719

JIm
